onnx/utils.py

```diff
@@ -21,3 +24,154 @@ def polish_model(model):  # type: (ModelProto) -> ModelProto
+        return nodes
+
+    def _collect_reachable_tensors(
+            self,
+            nodes,  # type: List[NodeProto]
+    ):  # type: (...) -> Tuple[List[TensorProto], List[ValueInfoProto]]
+        all_tensors_name = set()
+        for node in nodes:
+            for name in node.input:
+                all_tensors_name.add(name)
+            for name in node.output:
+                all_tensors_name.add(name)
+
+        initializer = [self.wmap[t] for t in self.wmap.keys() if t in all_tensors_name]
+        value_info = [self.vimap[t] for t in self.vimap.keys() if t in all_tensors_name]
+        assert(len(self.graph.sparse_initializer) == 0)
+        assert(len(self.graph.quantization_annotation) == 0)
+        return (initializer, value_info)
+
+    def _make_model(
+            self,
+            nodes,  # type: List[NodeProto]
+            inputs,  # type: List[ValueInfoProto]
+            outputs,  # type: List[ValueInfoProto]
+            initializer,  # type: List[TensorProto]
+            value_info  # type: List[ValueInfoProto]
+    ):  # type: (...) -> ModelProto
+        name = 'Extracted from {' + self.graph.name + '}'
+        graph = onnx.helper.make_graph(nodes, name, inputs, outputs, initializer=initializer,
+                                      value_info=value_info)
+
+        meta = {
+            'ir_version': self.model.ir_version,
+            'opset_imports': self.model.opset_import,
+            'producer_name': 'onnx.utils.extract_model',
+        }
+        return onnx.helper.make_model(graph, **meta)
+
+    def extract_model(
+            self,
+            input_names,  # type: List[Text]
+            output_names,  # type: List[Text]
+    ):  # type: (...) -> ModelProto
+        inputs = self._collect_new_inputs(input_names)
+        outputs = self._collect_new_outputs(output_names)
+        nodes = self._collect_reachable_nodes(input_names, output_names)
+        initializer, value_info = self._collect_reachable_tensors(nodes)
+        model = self._make_model(nodes, inputs, outputs, initializer, value_info)
+
+        return model
+
+
+def extract_model(
+        input_path,  # type: Text
+        output_path,  # type: Text
+        input_names,  # type: List[Text]
+        output_names  # type: List[Text]
+):  # type: (...) -> None
+    """Extracts sub-model from an ONNX model.
+
+    The sub-model is defined by the names of the input and output tensors *exactly*.
+
+    Note: For control-flow operators, e.g. If and Loop, the _boundary of sub-model_,
+    which is defined by the input and output tensors, should not _cut through_ the
+    subgraph that is connected to the _main graph_ as attributes of these operators.
+
+    Arguments:
+        input_path (string): The path to original ONNX model.
+        output_path (string): The path to save the extracted ONNX model.
+        input_names (list of string): The names of the input tensors that to be extracted.
```

Comment:
Looks like when input_names are not provided the the extracted sub graph uses the original inputs. This should be documented.

Reply:
Original input/output names are used in the _first commit_, and the current ([change](https://github.com/onnx/onnx/pull/2994/commits/dfe238f82e10f2a1c294841838535747651f851a?file-filters%5B%5D=.py#diff-498e239a2d86d6bcf902378224869c79L114)) implementation simply takes them as they are - empty is empty.

Reply:
input names being empty is not valid right? Then shouldn't there be a check for this on line 167?

Reply:
It could be surprising but an empty input name is valid actually, though not very meaningful in practice... Empty input names are cases that the extracted model is using _constant inputs_ i.e. initializers. @askhade 

The example below is a `Transpose` node with constant input, and we are extracting the model itself - as it has only one node.
```py
import onnx
from onnx import helper, TensorProto

# generate original ONNX model
M, N, K = 2, 3, 4

wdata = [i for i in range(N*K)]
weight = helper.make_tensor('weight', TensorProto.FLOAT, [N, K], wdata)

output = helper.make_tensor_value_info('output', TensorProto.FLOAT, [K, N])

transpose = helper.make_node('Transpose', ['weight'], ['output'], name='transpose', perm=(1, 0))

graph = helper.make_graph([transpose, ], 'demo', [], [output], [weight])
opset = helper.make_operatorsetid(onnx.defs.ONNX_DOMAIN, 9)
model = helper.make_model(graph, producer_name='toy-demo', ir_version=6, opset_imports=[opset])

onnx.checker.check_model(model)
onnx.save_model(model, 'example.onnx')

# extract model
onnx.utils.extract_model('example.onnx', 'extracted.onnx', list(), ['output'])
```